src/hir/mod.rs

```diff
@@ -12,8 +13,22 @@ pub use hir::*;
 /// This function leverages `crate::syntax::parse` and `translator::Translator::translate`
 /// to hide away most of the complexity of `bulloak`'s internal compiler.
 pub fn translate(tree: &str) -> anyhow::Result<Hir> {
-    let ast = crate::syntax::parse(tree)?;
+    let asts = crate::syntax::parse(tree)?;
+    let combiner = combiner::Combiner::new();
+    match combiner.verify(&asts) {
+        Err(_e) => {
+            // @follow-up - propagate this error
+        }
+        _ => {}
+    }
+    let mut hirs = Vec::new();
     let mut discoverer = crate::scaffold::modifiers::ModifierDiscoverer::new();
-    let modifiers = discoverer.discover(&ast);
-    Ok(translator::Translator::new().translate(&ast, modifiers))
+    let translator = translator::Translator::new();
+    for ast in &asts {
+        // @follow-up - use `map` instead of `for _ in`, here and elsewhere?
```

Comment:
It's fine to use for loops. Iterators are usually faster, but `bulloak` doesn't need to be super performant and a for loop is more readable oftentimes.